main-command/src/main/scala/sbt/internal/client/NetworkClient.scala

```diff
@@ -190,82 +378,497 @@ class NetworkClient(configuration: xsbti.AppConfiguration, arguments: List[Strin
+        errorStream.println(s"Caught exception writing command to server: $e")
+        running.set(false)
     }
-    lock.synchronized {
-      status.set("Processing")
+  }
+  def sendJson(method: String, params: String): String = {
+    val uuid = UUID.randomUUID.toString
+    val msg = s"""{ "jsonrpc": "2.0", "id": "$uuid", "method": "$method", "params": $params }"""
+    connection.sendString(msg)
+    uuid
+  }
+
+  def sendNotification(method: String, params: String): Unit = {
+    connection.sendString(s"""{ "jsonrpc": "2.0", "method": "$method", "params": $params }""")
+  }
+
+  private[this] class RawInputThread extends Thread("sbt-read-input-thread") with AutoCloseable {
+    setDaemon(true)
+    start()
+    val stopped = new AtomicBoolean(false)
+    val lock = new Object
+    override final def run(): Unit = {
+      @tailrec def read(): Unit = {
+        inputStream.read match {
+          case -1 =>
+          case b =>
+            lock.synchronized(stdinBytes.offer(b))
+            if (attached.get()) drain()
+            if (!stopped.get()) read()
+        }
+      }
+      try Terminal.console.withRawSystemIn(read())
+      catch { case _: InterruptedException | _: ClosedChannelException => stopped.set(true) }
+    }
+
+    def drain(): Unit = lock.synchronized {
+      while (!stdinBytes.isEmpty) {
+        val byte = stdinBytes.poll()
+        sendNotification(systemIn, byte.toString)
+      }
+    }
+
+    override def close(): Unit = {
+      RawInputThread.this.interrupt()
+    }
+  }
+
+  // copied from Aggregation
+  private def timing(startTime: Long, endTime: Long): String = {
+    import java.text.DateFormat
+    val format = DateFormat.getDateTimeInstance(DateFormat.MEDIUM, DateFormat.MEDIUM)
+    val nowString = format.format(new java.util.Date(endTime))
+    val total = (endTime - startTime + 500) / 1000
+    val totalString = s"$total s" +
+      (if (total <= 60) ""
+       else {
+         val maybeHours = total / 3600 match {
+           case 0 => ""
+           case h => f"$h%02d:"
+         }
+         val mins = f"${total % 3600 / 60}%02d"
+         val secs = f"${total % 60}%02d"
+         s" ($maybeHours$mins:$secs)"
+       })
+    s"Total time: $totalString, completed $nowString"
+  }
+  override def close(): Unit =
+    try {
+      running.set(false)
+      stdinBytes.offer(-1)
+      val mainThread = interactiveThread.getAndSet(null)
+      if (mainThread != null && mainThread != Thread.currentThread) mainThread.interrupt
+      connection.shutdown()
+      Option(inputThread.get).foreach(_.interrupt())
+    } catch {
+      case t: Throwable => t.printStackTrace(); throw t
+    }
+}
+class NetworkClient(
+    override val arguments: Arguments,
+    override val inputStream: InputStream,
+    override val errorStream: PrintStream,
+    override val printStream: PrintStream,
+) extends {
+  override val console: ConsoleInterface = {
+    val appender = ConsoleAppender("thin")
```

Comment:
Wouldn't this wire to the `printStream`?
```suggestion
    val appender = ConsoleAppender("thin", sbt.internal.util.ConsoleOut.printStreamOut(printStream))
```

Reply:
Good catch.